src/main.rs

```diff
@@ -0,0 +1,222 @@
+use std::time::Duration;
+
+use clap::Parser;
+
+use consul_external_dns::hetzner_dns;
+use tokio::time::{interval, sleep, MissedTickBehavior};
+
+use consul_external_dns::config::{Config, DnsProvider};
+use consul_external_dns::consul::ConsulClient;
+use consul_external_dns::dns_trait::DnsProviderTrait;
+use tokio_util::sync::CancellationToken;
+use uuid::Uuid;
+
+#[tokio::main]
+async fn main() {
+    let cancel = CancellationToken::new();
+    println!("Starting up Consul External DNS");
+
+    println!("=> Parsing configuration");
+    let config = Config::try_parse().expect("===> failed to parse configuration");
+    println!("===> parsed successfully");
+
+    let dns_provider: Box<dyn DnsProviderTrait> = match config.clone().dns_provider {
+        DnsProvider::Hetzner(config) => Box::new(hetzner_dns::HetznerDns { config }),
+    };
+
+    // Initialize Consul Client
+    println!("=> Creating Consul client");
+    let mut consul_client = create_consul_client(&config).await;
+    println!("===> created Consul client successfully");
+
+    // Create Consul session
+    println!("=> Creating Consul session");
+    let session_id = consul_client
+        .create_session()
+        .await
+        .expect("===> failed to create consul session");
+    println!("===> created Consul session successfully");
+
+    // Spawn a task to renew the session periodically
+    {
+        tokio::spawn(renew_session_periodically(
+            consul_client.clone(),
+            session_id,
+            cancel.clone(),
+        ));
+    }
```

Comment:
While this works, I would recommend something like the structure I used [here](https://github.com/svenstaro/wiresmith/commit/3bc7d727623cd863e8830c961bc966992ecda584) instead, because it feels structurally much cleaner to me.

Reply:
fixed